core-bundle/src/Migration/AbstractRecordedMigration.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Migration;
+
+use Contao\CoreBundle\Entity\Migration as MigrationEntity;;
+use Contao\CoreBundle\Repository\MigrationRepository;
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Contracts\Service\ServiceSubscriberTrait;
+
+abstract class AbstractRecordedMigration extends AbstractMigration
+{
+    use ServiceSubscriberTrait;
```

Comment:
I wonder if we should rather use a static service subscriber method here. :thinking: 
We're currently not using this anywhere else in the code base, do we?

I've been using this pattern quite heavily lately but I'm not sure I really like it. It makes you introduce traits for everything and then you'll end up with a lot of magic and more named things that need to be managed.

Reply:
> We're currently not using this anywhere else in the code base, do we?

No, but the Symfony documentation told me about this and this is the first time I have seen it and I wanted to try it 😁

Reply:
> I've been using this pattern quite heavily lately but I'm not sure I really like it. It makes you introduce traits for everything and then you'll end up with a lot of magic and more named things that need to be managed.

What intrigued me is that using it this way automatically defines the correct type - compared to `$this->container->get('…')`.

Reply:
Yes, but you'll get the same experience with just the methods (and their return types) + one additional `getSubscribedServices`. I did this in the image studio classes for example.

Reply:
This would also force you to set `autowiring: true` for your migration service. And then you could just use constructor injection as well which IMO is the easiest to understand. The lazyness could be nice on its own though. Ah, not easy. :slightly_smiling_face:

Reply:
Since we have a new abstract class anyway, then using a constructor argument to inject both services would be fine. I went this route because I originally extended the existing `AbstractMigration` class and there I did not want to introduce a new constructor.

Reply:
> This would also force you to set `autowiring: true` for your migration service.

Well no, actually all you need is

```yml
_instanceof:
    Contao\CoreBundle\Migration\AbstractRecordedMigration:
        calls:
            - [setContainer, ['@Psr\Container\ContainerInterface']]
```

Reply:
Isn't that more complex to setup or would you ship that in the ME?

Reply:
The Managed Edition already has autowiring enabled for the `App` namespace. I am not sure we really should introduce a constructor, otherwise you will always have to inject the same arguments for your own constructor and then passt it on to the parent? Which I find more cumbersome. The current solution is more in line with the `AbstractController` from Symfony for example, is it not?